arduino/smartcar/smartcar.ino

```diff
@@ -1,83 +1,200 @@
   mqtt.begin("127.0.0.1", 1883, net);
 
-  //print . while arduino is not connected to car
-  while(!mqtt.connect("SmartCarMQTT", "SmartCarMQTT", " ")) {
+  // print . while arduino is not connected to car
+  while (!mqtt.connect("SmartCarMQTT", "SmartCarMQTT", " ")) {
     Serial.println("MQTT Connecting...");
     delay(1000);
   }
-  if(mqtt.connected()) {
+  if (mqtt.connected()) {
     Serial.println("Connected to MQTT Broker");
   }
 
-  //subscribe to main topic w/ wildcard attached
-  mqtt.subscribe(THROTTLE_TOPIC, 1);
-  //on specific topics, it will do certain things
-  mqtt.onMessage([](String topic, String message){
-    Serial.println(topic + ": " + message);
-
-    if(topic == "/smartcar/controls/throttle") {
-      speed = message;
+  // subscribe to main topic w/ wildcard attached
+  mqtt.subscribe(MAINMQTT_TOPIC + "#", 1);
+  // on specific topics, it will do certain things
+  mqtt.onMessage([](String topic, String message) {
+    if (topic == THROTTLE_TOPIC) {
+      car.setSpeed(message.toInt());
+      // enter commands interpret received commands
+    } else if (topic == STEERING_TOPIC) {
+      car.setAngle(message.toInt());
+    } else if (topic == ESTOP_TOPIC) {
+      car.setSpeed(0);
+      mqtt.publish(ESTOP_TOPIC, "Emergency Stop. Speed has been set to zero.");
+      Serial.println("Car has stopped");
+    } else if (topic == BLINKERS_TOPIC) {
+      if (message == "left") {
+        Serial.println("Blinker to left");
+        blinkerStatus = "left";
+      } else if (message == "right") {
+        Serial.println("Blinker to right");
+        blinkerStatus = "right";
+      } else if (message == "off") {
+        blinkerStatus = "off";
+      }
     }
   });
-
 }
 
 void loop() {
-  //when connected, keep checking for messages
+  gyroscope.update();
+  // when connected, keep checking for messages
   if (mqtt.connected()) {
     mqtt.loop();
-    //delay to not overload the CPU
-    delay(100);
-    mqtt.publish("/smartcar/derivedData/speed", speed);
+    // delay to not overload the CPU
+    delay(1);
+
+    const auto currentTime = millis();
+
+    /**
+     * @brief if the it has been 1 second since the last transmission
+     */
+    static auto previousTransmission = 0UL;
+    static auto previousHeading = 0;
+    if (currentTime - previousTransmission >= 1000UL) {
+      // Publishing IR sensor
+      const auto ir_distance = String(frontIR.getDistance());
+      mqtt.publish(IR_TOPIC, ir_distance);
+
+      // Publishing average distance travelled via both Odometers
+      //  (distance logged is in centimeters)
+      const auto avg_distance = String(
+          (leftOdometer.getDistance() + rightOdometer.getDistance()) / 2);
+      mqtt.publish(ODOMETER_DISTANCE, avg_distance);
+
+      // Publishing average speed of the car via both Odometers
+      //  (speed logged is in meters per second)
+      const auto avg_speed =
+          String((leftOdometer.getSpeed() + rightOdometer.getSpeed()) / 2);
```

Comment:
If you have a `SmartCar` you can instead do `car.getSpeed()`)

Reply:
Thanks, we will do!
btw, is there a way to get the throttles percentage? We couldn't find anything in the docs for BrushedMotor.

Reply:
No, if you want to keep track of the speed you have set you can... keep track of it yourself. :stuck_out_tongue: